src/systems/filecoin_markets/storage_market/storage_market_actor.go

```diff
@@ -242,74 +235,299 @@ func (a *StorageMarketActorCode_I) PublishStorageDeals(rt Runtime, newStorageDea
 
-	return lastDealExpiration
+	return initialUtilizationInfo
+}
+
+// unlock remaining payments and return all UnlockedStorageFee to Provider
+// remove deals from ActiveDeals
+// return collaterals to both miner and client
+func (st *StorageMarketActorState_I) _expireStorageDeals(rt Runtime, dealIDs []deal.DealID, lastChallengeEndEpoch block.ChainEpoch) {
+	for _, dealID := range dealIDs {
+
+		newExpiredDeal, found := st.ActiveDeals()[dealID]
+
+		if !found {
+			// rt.Abort("sma._expireStorageDeals: dealID not found")
+			return
+		}
+
+		duration := lastChallengeEndEpoch - newExpiredDeal.LastPaymentEpoch()
+		if duration > 0 {
+			fee := actor.TokenAmount(uint64(duration) * uint64(newExpiredDeal.Deal().Proposal().StoragePricePerEpoch()))
+
+			if fee > newExpiredDeal.LockedStorageFee() {
+				// rt.Abort("sma._expireStorageDeals: cannot unlock more than already locked.")
+				return
+			}
+
+			newExpiredDeal.Impl().LastPaymentEpoch_ = lastChallengeEndEpoch
+			newExpiredDeal.Impl().UnlockStorageFee(fee)
+		}
+
+		delete(st.ActiveDeals(), dealID)
+
+		// should we check if LockedStorageFee is now 0?
+		// credit UnlockedStorageFee to miner
+		st._unlockBalance(rt, newExpiredDeal.Deal().Proposal().Provider(), newExpiredDeal.UnlockedStorageFee())
+
+		// return storage deal collaterals to both miners and client
+		st._unlockBalance(rt, newExpiredDeal.Deal().Proposal().Provider(), newExpiredDeal.ProviderCollateralRemaining())
+		st._unlockBalance(rt, newExpiredDeal.Deal().Proposal().Client(), newExpiredDeal.Deal().Proposal().TotalClientCollateral())
+
+	}
+}
+
+func (st *StorageMarketActorState_I) _creditStorageDeals(rt Runtime, dealIDs []deal.DealID, lastChallengeEndEpoch block.ChainEpoch) {
+
+	for _, dealID := range dealIDs {
+		activeDeal, found := st.ActiveDeals()[dealID]
+
+		if !found {
+			rt.Abort("sma._creditStorageDeals: dealID not found.")
+		}
+
+		duration := lastChallengeEndEpoch - activeDeal.LastPaymentEpoch()
+
+		if duration <= 0 {
+			// return for this particular deal instead of aborting the whole operation
+			// rt.Abort("sma._creditStorageDeals: no new payment to be credited.")
+			return
+		}
+
+		fee := actor.TokenAmount(uint64(duration) * uint64(activeDeal.Deal().Proposal().StoragePricePerEpoch()))
+
+		if fee > activeDeal.LockedStorageFee() {
+			// rt.Abort("sma._creditStorageDeals: cannot unlock more than already locked.")
+			return
+		}
+
+		activeDeal.Impl().LastPaymentEpoch_ = lastChallengeEndEpoch
+
+		// potentially unnecessary as we can unlock funds for provider directly
+		// unless the protocol plans on refunding client for MaxFaultCount consecutive fails
+		activeDeal.Impl().UnlockStorageFee(fee)
+
+		// TODO: align on provider storage fee withdrawal
+	}
+
+}
+
+func (st *StorageMarketActorState_I) _slashDeclaredFaults(rt Runtime, dealIDs []deal.DealID) {
+
+	for _, dealID := range dealIDs {
+
+		deal, found := st.ActiveDeals()[dealID]
+
+		if !found {
+			// move on to the next deal and keep slashing
```

Comment:
this might be an error condition in some design versions. im ok with no error, but make it clearer -- say why, eg "the deal may no longer be active, and we are OK with a call to this function referencing inactive deals" or something like that.

Reply:
agree